client/acl.go

```diff
@@ -227,3 +255,116 @@ func (c *Client) resolvePolicies(secretID string, policies []string) ([]*structs
+		policyNames []string
+
+		// missingRoleIDs are the roles linked which are not found within our
+		// cache. These must be looked up from the server via and RPC, so we
+		// can correctly identify the policy links.
+		missingRoleIDs []string
+
+		// expiredRoleIDs are the roles linked which have been found within our
+		// cache, but are expired. These must be looked up from the server via
+		// and RPC, so we can correctly identify the policy links.
+		expiredRoleIDs []string
+	)
+
+	for _, roleLink := range roleLinks {
+
+		// Look within the cache to see if the role is already present. If we
+		// do not find it, add the ID to our tracking, so we look this up via
+		// RPC.
+		raw, ok := c.roleCache.Get(roleLink.ID)
+		if !ok {
+			missingRoleIDs = append(missingRoleIDs, roleLink.ID)
+			continue
+		}
+
+		// If the cached value is expired, add the ID to our tracking, so we
+		// look this up via RPC. Otherwise, iterate the policy links and add
+		// each policy name to our return object tracking.
+		cached := raw.(*cachedACLValue)
+		if cached.Age() <= c.GetConfig().ACLRoleTTL {
+			for _, policyLink := range cached.Role.Policies {
+				policyNames = append(policyNames, policyLink.Name)
+			}
+		} else {
+			expiredRoleIDs = append(expiredRoleIDs, cached.Role.ID)
+		}
+	}
+
+	// Hot-path: we were able to resolve all ACL roles via the cache and
+	// generate a list of linked policy names. Therefore, we can avoid making
+	// any RPC calls.
+	if len(missingRoleIDs)+len(expiredRoleIDs) == 0 {
+		return policyNames, nil
+	}
+
+	// Created a combined list of role IDs that we need to lookup from server
+	// state.
+	roleIDsToFetch := missingRoleIDs
+	roleIDsToFetch = append(roleIDsToFetch, expiredRoleIDs...)
+
+	// Generate an RPC request to detail all the ACL roles that we did not find
+	// or were expired within the cache.
+	roleByIDReq := structs.ACLRolesByIDRequest{
+		ACLRoleIDs: roleIDsToFetch,
+		QueryOptions: structs.QueryOptions{
+			Region:     c.Region(),
+			AuthToken:  secretID,
+			AllowStale: true,
+		},
+	}
+
+	var roleByIDResp structs.ACLRolesByIDResponse
+
+	// Perform the RPC call to detail the required ACL roles. If the RPC call
+	// fails, and we are only updating expired cache entries, use the expired
+	// entries. This allows use to handle intermittent failures.
+	err := c.RPC(structs.ACLGetRolesByIDRPCMethod, &roleByIDReq, &roleByIDResp)
+	if err != nil {
+		if len(missingRoleIDs) == 0 {
+			c.logger.Warn("failed to resolve ACL roles, using expired cached value", "error", err)
+			for _, aclRole := range roleByIDResp.ACLRoles {
+				for _, rolePolicyLink := range aclRole.Policies {
+					policyNames = append(policyNames, rolePolicyLink.Name)
+				}
+			}
+			return policyNames, nil
+		}
+		return nil, err
+	}
+
+	// Generate a timestamp for the cache entry. We do not need to use a
+	// timestamp per ACL role response integration.
+	now := time.Now()
+
+	for _, aclRole := range roleByIDResp.ACLRoles {
```

Comment:
If the RPC didn't return a role for a given ID, it won't be in this map. If we get to the end of this loop and there are still missing roles, should we be returning an error or otherwise rejecting the operation so that we don't accidentally "fail open"?

Reply:
I think the current behaviour wouldn't fail open; I am going to try and walk through this here so it becomes clearer in my head.

The timing situation here could be that a developer makes a client API call just before a cluster operator deletes an ACL role. The Raft deletion occurs before we lookup the role from the server cluster, this lookup happens because either the cache entry is old, or it does not exist. The API call is therefore made using a token which contains a number of roles and policies; one or more of these ACL roles do not exist and therefore are not available within this loop.

The return object and purpose of this function is to generate an array of policy names which are linked to the token by ACL roles. Therefore, the policies linked from a missing role will not be added to the array. This is OK if they are policies which "allow" actions as the ACL object which is complied later would not have these permissions. If the policy was a "deny", this would still be OK as the role has been deleted removing this restriction.

The function has handling for when the RPC call errors to protect from "fail open" and will only fallback to the cache if we can rely on it for all ACL roles. The RPC succeeding means we were able to perform a state lookup for the roles, and therefore any missing roles would be missing from state and considered deleted.

Reply:
>  If the policy was a "deny", this would still be OK as the role has been deleted removing this restriction.

Yeah I think what I was missing here was as you mentioned in the other PR, that a token can't be given a role that doesn't exist in the first place. So we can only hit this problem on the delete path as you're saying here. LGTM :+1